### ISSUE TITLE
Possible to infer snapshot name from test name?

### ISSUE THREAD

**Issue by JoshMcguigan** [mitsuhiko/insta#24]
I have a pattern of naming my snapshots the same as the test name. Would it be possible for insta to infer the snapshot name from the name of the test, perhaps while still allowing a user to optionally pass in a different name if they'd like?

**Comment by mitsuhiko** [comment]
Sadly the only reasonable way to make this work would need this: https://github.com/rust-lang/rfcs/issues/1743

**Comment by JoshMcguigan** [comment]
I think it could be done with procedural macros? 

```rust
// current usage
#[test]
fn test_snapshots() {
    let value = vec![1, 2, 3];
    assert_debug_snapshot_matches!("snapshot_name", value);
}

// proposed
#[assert_debug_snapshot_matches]
fn test_snapshots() -> Impl Debug {
    vec![1, 2, 3]
}
```

I'm not sure, but it may even be possible to determine which type of snapshot test to use based on the return type of the function.

```rust
// maybe possible??
#[snapshot]
fn test_snapshots() -> Impl Debug {
    vec![1, 2, 3]
}
```

**Comment by mitsuhiko** [comment]
I found a way to make it work. Rust test names the threads so I can access the thread name to get this information. Implemented in #27.
